File: admin/src/api/file-attachment.ts
import { http, type ApiEnvelope } from '../utils/request'

import type { PaginatedData } from './admin-user'

export interface FileUploaderSummary {
  id: number
  username: string
  display_name: string
}

export interface FileItem {
  id: number
  original_name: string
  mime_type: string
  extension: string
  size: number
  url: string
  uploader: FileUploaderSummary | null
  created_at: string
}

export interface FileReferenceItem {
  id: number
  file_id: number
  ref_type: string
  ref_id: string
  ref_name: string | null
  ref_path: string | null
  created_at: string
}

export interface FileReferenceResponse {
  file_id: number
  reference_count: number
  references: FileReferenceItem[]
}

export interface FileDetailResponse extends FileItem {
  storage_driver: string
  checksum: string
  reference_count: number
  references: FileReferenceItem[]
  download_url: string
  can_delete: boolean
}

export interface FileUploadResponse {
  id: number
  original_name: string
  mime_type: string
  size: number
  url: string
  created_at: string
}

export interface FileListQuery {
  page?: number
  per_page?: number
  keyword?: string
  mime_type?: string
  uploader_id?: number
  date_from?: string
  date_to?: string
}

export async function getFiles(query?: FileListQuery) {
  const response = await http.get<ApiEnvelope<PaginatedData<FileItem>>>('/files', { params: query })
  return response.data.data
}

export async function getFileDetail(id: number) {
  const response = await http.get<ApiEnvelope<FileDetailResponse>>(`/files/${id}`)
  return response.data.data
}

export async function getFileReferences(id: number) {
  const response = await http.get<ApiEnvelope<FileReferenceResponse>>(`/files/${id}/references`)
  return response.data.data
}

export async function uploadFile(file: File) {
  const formData = new FormData()
  formData.append('file', file)
  const response = await http.post<ApiEnvelope<FileUploadResponse>>('/files/upload', formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  })
  return response.data.data
}

export async function deleteFile(id: number) {
  await http.delete<ApiEnvelope<null>>(`/files/${id}`)
}

export function getFileDownloadUrl(id: number) {
  return `/admin-api/files/${id}/download`
}

export async function downloadFile(id: number) {
  const response = await http.get<Blob>(`/files/${id}/download`, { responseType: 'blob' })
  return response.data
}

File: server/internal/admin/dto/file_attachment.go
package dto

import "time"

/**
 * FileListQuery 表示文件列表查询参数。
 */
type FileListQuery struct {
	Page       int    `form:"page" validate:"omitempty,min=1"`
	PerPage    int    `form:"per_page" validate:"omitempty,min=1,max=100"`
	Keyword    string `form:"keyword" validate:"omitempty,max=255"`
	MimeType   string `form:"mime_type" validate:"omitempty,max=128"`
	UploaderID uint64 `form:"uploader_id" validate:"omitempty"`
	DateFrom   string `form:"date_from" validate:"omitempty"`
	DateTo     string `form:"date_to" validate:"omitempty"`
}

/**
 * FileItem 表示文件列表项。
 */
type FileItem struct {
	ID           uint64               `json:"id"`
	OriginalName string               `json:"original_name"`
	MimeType     string               `json:"mime_type"`
	Extension    string               `json:"extension"`
	Size         uint64               `json:"size"`
	URL          string               `json:"url"`
	Uploader     *FileUploaderSummary `json:"uploader"`
	CreatedAt    time.Time            `json:"created_at"`
}

/**
 * FileReferenceItem 表示文件引用记录。
 */
type FileReferenceItem struct {
	ID        uint64    `json:"id"`
	FileID    uint64    `json:"file_id"`
	RefType   string    `json:"ref_type"`
	RefID     string    `json:"ref_id"`
	RefName   *string   `json:"ref_name"`
	RefPath   *string   `json:"ref_path"`
	CreatedAt time.Time `json:"created_at"`
}

/**
 * FileReferenceResponse 表示文件引用查询结果。
 */
type FileReferenceResponse struct {
	FileID         uint64              `json:"file_id"`
	ReferenceCount int64               `json:"reference_count"`
	References     []FileReferenceItem `json:"references"`
}

/**
 * FileDetailResponse 表示文件详情。
 */
type FileDetailResponse struct {
	FileItem
	StorageDriver  string              `json:"storage_driver"`
	Checksum       string              `json:"checksum"`
	ReferenceCount int64               `json:"reference_count"`
	References     []FileReferenceItem `json:"references"`
	DownloadURL    string              `json:"download_url"`
	CanDelete      bool                `json:"can_delete"`
}

/**
 * FileUploaderSummary 表示文件上传者摘要。
 */
type FileUploaderSummary struct {
	ID          uint64 `json:"id"`
	Username    string `json:"username"`
	DisplayName string `json:"display_name"`
}

/**
 * FileUploadResponse 表示文件上传成功响应。
 */
type FileUploadResponse struct {
	ID           uint64    `json:"id"`
	OriginalName string    `json:"original_name"`
	MimeType     string    `json:"mime_type"`
	Size         uint64    `json:"size"`
	URL          string    `json:"url"`
	CreatedAt    time.Time `json:"created_at"`
}
